sys/sys_mmap.c

```diff
@@ -0,0 +1,79 @@
+#include <sys_mmap.h>
+#include <thread.h>
+#include <stdc.h>
+#include <errno.h>
+#include <vm_map.h>
+#include <vm_pager.h>
+#include <rwlock.h>
+
+int sys_mmap(thread_t *td, syscall_args_t *args) {
+  vm_addr_t addr = args->args[0];
+  size_t length = args->args[1];
+  vm_prot_t prot = args->args[2];
+  int flags = args->args[3];
+
+  kprintf("[syscall] mmap(%p, %zu, %d, %d)\n", (void *)addr, length, prot,
+          flags);
+
+  int error = 0;
+  vm_addr_t result = do_mmap(addr, length, prot, flags, &error);
+  if (error < 0)
+    return -error;
+  return result;
+}
+
+vm_addr_t do_mmap(vm_addr_t addr, size_t length, vm_prot_t prot, int flags,
+                  int *error) {
+  thread_t *td = thread_self();
+  vm_map_t *vmap = td->td_uspace;
+
+  assert(vmap);
+
+  if (addr >= vmap->pmap->end) {
+    /* mmap cannot callocate memory in kernel space! */
+    *error = EINVAL;
+    return MMAP_FAILED;
+  }
+
+  if (!flags & MMAP_FLAG_ANONYMOUS) {
+    log("Non-anonymous memory mappings are not yet implemented.");
+    *error = EINVAL;
+    return MMAP_FAILED;
+  }
+
+  {
+    rw_scoped_enter(&vmap->rwlock, RW_WRITER);
+
+    length = roundup(length, PAGESIZE);
```

Comment:
Only acquire lock for actions that require it. Please move lines 47-54 outside the scope.